File: backend/api/people.py
"""Eklenen kişiler — eş, çocuk, ebeveyn, yakın (P-turu).

## Neden sunucudan yazılıyor

Kayıt `users/{uid}/people/{id}` altında ve Firestore kuralında **okuma
sahibe açık, yazma kapalı** (sohbet arşiviyle aynı desen). Sebep kontenjan:
kaç kişi eklenebileceği bir ürün gerçeğidir ve istemcide zorlanamaz —
Firestore kuralları koleksiyon sayamaz. Bu yüzden ekleme/silme buradan
geçer, okuma istemcinin kendi akışından.

## Yanıtta ne YOK

Kişinin adı. Sunucu onu hiç bilmiyor (bkz. services/people_service.py
dosya başı): etiket cihazda yaşıyor, burada yalnız `relation` var.
"""
from __future__ import annotations

import logging

from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel, Field

from core.auth import AuthUser, get_current_user
from core.i18n import get_language
from core.messages import text
from services import people_service

logger = logging.getLogger(__name__)
router = APIRouter()


class PersonRequest(BaseModel):
    """Kişinin doğum verisi. Ad alanı BİLİNÇLİ OLARAK YOK."""

    relation: str = Field(default="other", max_length=32)
    #: "YYYY-MM-DD"
    birth_date: str = Field(min_length=10, max_length=10)
    #: "HH:MM" ya da None/boş — **bilinmiyor** demektir, 12:00 varsayılmaz.
    birth_time: str | None = Field(default=None, max_length=5)
    birth_city: str = Field(min_length=1, max_length=120)
    #: ISO-2 ülke kodu (şehir seçiciden).
    birth_nation: str | None = Field(default=None, max_length=2)
    gender: str = Field(default="female", max_length=16)

    def to_fields(self) -> dict:
        """Firestore alan adlarına çevirir (doğrulama serviste)."""
        return {
            "relation": self.relation,
            "birthDate": self.birth_date,
            "birthTime": self.birth_time,
            "birthCity": self.birth_city,
            "birthNation": self.birth_nation,
            "gender": self.gender,
        }


def _liste(uid: str) -> dict:
    kisiler = people_service.list_people(uid)
    return {
        "people": kisiler,
        "used": len(kisiler),
        "limit": people_service.slot_limit(uid),
        "relations": list(people_service.RELATIONS),
    }


@router.get("")
def list_people(user: AuthUser = Depends(get_current_user)):
    """Kişiler + kontenjan durumu.

    İstemci listeyi Firestore akışından da okuyor; bu uç kontenjanı
    (`limit`) bildirmek için var — abonelik durumu istemcide anlık
    değişebiliyor ve "kaç kişi ekleyebilirim" sorusunun cevabı sunucunun.
    """
    return {"status": "success", "data": _liste(user.uid)}


@router.post("")
def create_person(req: PersonRequest,
                  user: AuthUser = Depends(get_current_user),
                  lang: str = Depends(get_language)):
    """Kişi ekler. Kontenjan doluysa 402 + ``X-Paywall-Reason: people``."""
    kayit = people_service.create_person(user.uid, req.to_fields(), lang=lang)
    return {"status": "success", "data": {"person": kayit, **_liste(user.uid)}}


@router.patch("/{person_id}")
def update_person(person_id: str, req: PersonRequest,
                  user: AuthUser = Depends(get_current_user),
                  lang: str = Depends(get_language)):
    """Doğum verisini düzeltir; Büyük Üçlü rozetleri birlikte tazelenir.

    Kontenjan BURADA sorulmaz: var olan bir kaydı düzeltmek yeni yer
    açmıyor. Aboneliği biten kullanıcı 10 kişisini düzeltmeye devam eder,
    yalnız yenisini ekleyemez — elindeki veriyi rehin almak olurdu.
    """
    kayit = people_service.update_person(user.uid, person_id,
                                         req.to_fields(), lang=lang)
    return {"status": "success", "data": {"person": kayit}}


@router.delete("/{person_id}")
def delete_person(person_id: str,
                  user: AuthUser = Depends(get_current_user),
                  lang: str = Depends(get_language)):
    if not people_service.delete_person(user.uid, person_id):
        raise HTTPException(status_code=404, detail=text("people.missing", lang))
    return {"status": "success", "data": _liste(user.uid)}

File: apps/mobile/lib/core/people.dart
/// EKLENEN KİŞİLER — kullanıcının kendi girdiği eş, çocuk, ebeveyn, yakın.
///
/// ## Arkadaştan farkı
///
/// Arkadaş karşılıklı rızayla eklenen bir Rytho KULLANICISIDIR; bu kişi
/// değildir. Hesabı yok, tepki gönderemez, serisi olmaz. Ekranda o
/// etkileşimleri çizmek olmayan bir şeyi varmış gibi göstermek olurdu
/// (bkz. features/friends/friends_screen.dart satır anatomisi).
///
/// ## Ad neden sunucuda değil
///
/// Bu kişilerin rızası ALINAMIYOR. Ad + doğum tarihi + doğum yeri üçlüsü
/// kimlik doğrulama sorularında kullanılan hassas bir kümedir; üçüncü bir
/// kişinin bu kümesini rızası olmadan sunucuya yazmıyoruz. Doğum verisi
/// sunucuda (harita orada hesaplanıyor), **etiket cihazda**. Sunucunun
/// gördüğü tek kimlik işareti [Person.relation] ve AI kişiyi "eşin" /
/// "çocuğun" diye anıyor.
///
/// Bedeli: telefon değişince doğum verisi durur, etiketler yeniden yazılır.
/// Bu bilinçli bir takas — bkz. backend/services/people_service.py.
library;

import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:dio/dio.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'package:shared_preferences/shared_preferences.dart';

import 'api.dart';
import 'providers.dart';

/// Kişi türleri. Backend `people_service.RELATIONS` ile **birebir aynı**
/// olmak zorunda; tür yalnız simge değil, ilişki eksenlerinin adını ve
/// AI'nın çerçevesini belirliyor (çocukla "çekim" ekseni gösterilmez).
const List<String> kRelations = [
  'partner',
  'child',
  'parent',
  'sibling',
  'friend',
  'work',
  'other',
];

class Person {
  const Person({
    required this.id,
    required this.relation,
    required this.birthDate,
    required this.birthCity,
    this.label,
    this.birthTime,
    this.birthNation,
    this.gender = 'female',
    this.sunSign,
    this.moonSign,
    this.ascendant,
  });

  final String id;
  final String relation;

  /// Kullanıcının verdiği ad — **yalnızca bu cihazda**. Sunucu bilmez.
  /// Boşsa arayüz ilişki türünün yerelleştirilmiş adını gösterir.
  final String? label;

  /// `YYYY-MM-DD`.
  final String birthDate;

  /// `HH:mm` ya da **null**: doğum saati bilinmiyor. Null "12:00 varsayalım"
  /// DEĞİLDİR — sunucu Yükselen/ev üretmez ve okumada bunu beyan eder
  /// (core/birth_record.dart'taki kuralın aynısı).
  final String? birthTime;

  final String birthCity;
  final String? birthNation;
  final String gender;

  final String? sunSign;
  final String? moonSign;
  final String? ascendant;

  bool get hourKnown => birthTime != null;

  factory Person.fromDoc(String id, Map<String, dynamic> data,
      {String? label}) {
    return Person(
      id: id,
      relation: data['relation'] as String? ?? 'other',
      label: label,
      birthDate: data['birthDate'] as String? ?? '',
      // Alanın YOKLUĞU saatin bilinmediği anlamına gelir; varsayılan konmaz.
      birthTime: data['birthTime'] as String?,
      birthCity: data['birthCity'] as String? ?? '',
      birthNation: data['birthNation'] as String?,
      gender: data['gender'] as String? ?? 'female',
      sunSign: data['sunSign'] as String?,
      moonSign: data['moonSign'] as String?,
      ascendant: data['ascendant'] as String?,
    );
  }

  /// Rapor uçlarının beklediği gövde. `birthPayload` ile aynı sözleşme
  /// (core/api.dart) — kişinin ADI gönderilmez, çünkü sunucu prompt'ta
  /// ilişki etiketini kullanıyor.
  Map<String, dynamic> toBirthPayload() {
    final tarih = birthDate.split('-').map(int.parse).toList();
    final saat = (birthTime ?? '12:00').split(':').map(int.parse).toList();
    return {
      'year': tarih[0],
      'month': tarih[1],
      'day': tarih[2],
      'hour': saat[0],
      'minute': saat[1],
      'hour_known': birthTime != null,
      'city': birthCity,
      'nation': birthNation,
      'gender': gender,
    };
  }
}

FirebaseFirestore get _db => FirebaseFirestore.instance;

CollectionReference<Map<String, dynamic>> _peopleOf(String uid) =>
    _db.collection('users').doc(uid).collection('people');

// ---------------------------------------------------------------------------
// Etiketler — CİHAZDA
// ---------------------------------------------------------------------------

String _labelKey(String personId) => 'person-label-$personId';

Future<String?> personLabel(String personId) async {
  final prefs = await SharedPreferences.getInstance();
  return prefs.getString(_labelKey(personId));
}

Future<void> setPersonLabel(String personId, String label) async {
  final prefs = await SharedPreferences.getInstance();
  final temiz = label.trim();
  if (temiz.isEmpty) {
    await prefs.remove(_labelKey(personId));
  } else {
    await prefs.setString(_labelKey(personId), temiz);
  }
}

Future<void> _forgetLabel(String personId) async {
  final prefs = await SharedPreferences.getInstance();
  await prefs.remove(_labelKey(personId));
}

/// Türkçe-güvenli küçük harf: `toLowerCase` "İ"yi `i + birleşik nokta`
/// yapar ve `contains` sessizce kaçırır (bkz. backend
/// `prompt_composer.normalize` — aynı tuzağın istemci tarafı).
/// GT-turu'nda dışa açıldı: @-bahsetme süzgeci de aynı kuralı kullanır.
String normalizeTr(String s) => s.toLowerCase().replaceAll('̇', '');

String _normalize(String s) => normalizeTr(s);

/// KA6: mesaj cihazdaki kişi etiketlerinden birini (adını) anıyorsa o
/// kişinin kimliği; yoksa null.
///
/// Ad SUNUCUYA GİTMEZ — eşleşme cihazda yapılır, yalnız `person_id`
/// gönderilir; sunucu kimliği ilişki türü fısıltısına çevirir. Böylece
/// kullanıcı ana sekmeden "Ayşe'yle aram nasıl?" diye sorduğunda da doğru
/// kişinin ölçümü sohbete girer. Kısa etiketler (<3 harf) yanlış pozitif
/// riskine karşı eşleşmeye girmez.
String? matchPersonIdByLabel(String message, List<Person> people) {
  final metin = _normalize(message);
  for (final kisi in people) {
    final etiket = kisi.label?.trim();
    if (etiket == null || etiket.length < 3) continue;
    if (metin.contains(_normalize(etiket))) return kisi.id;
  }
  return null;
}

// ---------------------------------------------------------------------------
// Yazma — SUNUCUDAN (kontenjan istemcide zorlanamaz)
// ---------------------------------------------------------------------------

Map<String, dynamic> _body(Person kisi) => {
      'relation': kisi.relation,
      'birth_date': kisi.birthDate,
      'birth_time': kisi.birthTime,
      'birth_city': kisi.birthCity,
      'birth_nation': kisi.birthNation,
      'gender': kisi.gender,
    };

/// Kişi ekler ve etiketi CİHAZA yazar. Kontenjan doluysa sunucu 402
/// (`X-Paywall-Reason: people`) döner ve istisna çağırana çıkar.
///
/// Sıra önemli: kimlik sunucudan gelir, etiket ancak ondan sonra
/// yazılabilir. Etiket yazımı düşerse kayıt yine de vardır — arayüz
/// ilişki türünün adını gösterir, veri kaybı olmaz.
Future<String> createPerson(Dio dio, Person kisi, {String? label}) async {
  final response = await dio.post('/api/v1/people', data: _body(kisi));
  final data = Map<String, dynamic>.from(response.data['data'] as Map);
  final id = (data['person'] as Map)['id'] as String;
  if (label != null && label.trim().isNotEmpty) {
    await setPersonLabel(id, label);
  }
  return id;
}

Future<void> updatePerson(Dio dio, Person kisi, {String? label}) async {
  await dio.patch('/api/v1/people/${kisi.id}', data: _body(kisi));
  if (label != null) await setPersonLabel(kisi.id, label);
}

Future<void> deletePerson(Dio dio, String personId) async {
  await dio.delete('/api/v1/people/$personId');
  // Etiket sunucu silmeyi onayladıktan SONRA silinir: uç düşerse kayıt
  // duruyor demektir ve etiketi önden silmek kişiyi adsız bırakırdı.
  await _forgetLabel(personId);
}

// ---------------------------------------------------------------------------
// Sağlayıcılar
// ---------------------------------------------------------------------------

/// Eklenen kişiler — Firestore akışı + cihazdaki etiketler.
final peopleProvider = StreamProvider<List<Person>>((ref) {
  final user = ref.watch(authStateProvider).value;
  if (user == null) return Stream.value(const <Person>[]);

  return _peopleOf(user.uid)
      .snapshots()
      .asyncMap((snapshot) async {
    final kisiler = <Person>[];
    for (final doc in snapshot.docs) {
      kisiler.add(Person.fromDoc(doc.id, doc.data(),
          label: await personLabel(doc.id)));
    }
    // Tür sırası: eş, çocuk, ebeveyn... (kRelations sırası) — kullanıcının
    // hayatındaki yakınlık sırasına en yakın düzen.
    kisiler.sort((a, b) {
      final byRelation = kRelations.indexOf(a.relation)
          .compareTo(kRelations.indexOf(b.relation));
      if (byRelation != 0) return byRelation;
      return (a.label ?? '').compareTo(b.label ?? '');
    });
    return kisiler;
  });
});

/// Kontenjan durumu: kaç kişi eklenmiş, kaç kişiye izin var.
///
/// Sunucudan gelir çünkü sınır abonelik durumuna bağlı ve "kaç kişi
/// ekleyebilirim" sorusunun cevabı sunucunundur.
final personSlotsProvider =
    FutureProvider<({int used, int limit})>((ref) async {
  final user = ref.watch(authStateProvider).value;
  if (user == null) return (used: 0, limit: 0);
  // Kişi eklenince/silinince tazelensin.
  ref.watch(peopleProvider);
  final dio = ref.watch(apiProvider);
  final response = await dio.get('/api/v1/people');
  final data = Map<String, dynamic>.from(response.data['data'] as Map);
  return (
    used: (data['used'] as num?)?.toInt() ?? 0,
    limit: (data['limit'] as num?)?.toInt() ?? 0,
  );
});
